### DISCUSSION TITLE
Confused about code locations and kubernetes

### DISCUSSION THREAD

**Discussion by the4thamigo-uk** [dagster-io/dagster#18799] [Q&A] [answered]
I have been experimenting with dagster locally and everything works great. I now want to move my pipelines to kubernetes. I have installed the main helm chart, and I have separately deployed a code location with a grpc endpoint. The code location contains a trivial single asset that returns 'Hello'. I have connected dagster  to the code location, using the values.yml (workspace/servers), and in the UI I can see the asset. However when I try to manually materialize it, I am getting the below error.

I was naively thinking that the asset would be materialized in the code location pod, via an instruction sent over the grpc interface. But it looks like it is trying to spin up a new pod, as it is using the k8s client and seems to be compaining that it doesnt know which image to use.

I have read through the docs, but it is quite confusing...

I have two basic questions, which I think would get me along the path to understanding how this all works : 

1. How do I configure my dagster deployment and code location deployment so that the asset is materialized in the code location deployment, without spinning up new pods.
2. How do I configure my dagster deployment and code location deployment so that the asset is materialized in a new k8s pod (i.e. fixing the error I am currently seeing).

```
dagster_k8s.client.DagsterK8sUnrecoverableAPIError: Unexpected error encountered in Kubernetes API Client.
  File "/usr/local/lib/python3.10/site-packages/dagster/_daemon/run_coordinator/queued_run_coordinator_daemon.py", line 346, in _dequeue_run
    instance.run_launcher.launch_run(LaunchRunContext(dagster_run=run, workspace=workspace))
  File "/usr/local/lib/python3.10/site-packages/dagster_k8s/launcher.py", line 282, in launch_run
    self._launch_k8s_job_with_args(job_name, args, run)
  File "/usr/local/lib/python3.10/site-packages/dagster_k8s/launcher.py", line 263, in _launch_k8s_job_with_args
    self._api_client.create_namespaced_job_with_retries(body=job, namespace=namespace)
  File "/usr/local/lib/python3.10/site-packages/dagster_k8s/client.py", line 833, in create_namespaced_job_with_retries
    k8s_api_retry_creation_mutation(
  File "/usr/local/lib/python3.10/site-packages/dagster_k8s/client.py", line 183, in k8s_api_retry_creation_mutation
    raise DagsterK8sUnrecoverableAPIError(
The above exception was caused by the following exception:
kubernetes.client.exceptions.ApiException: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Audit-Id': '38e5ea20-9b7a-4cb1-8517-75bf123162e0', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '273d3580-efaa-429a-be45-af4bf554bea2', 'X-Kubernetes-Pf-Prioritylevel-Uid': '478a6c7e-1d28-4317-8662-42723a1e34c0', 'Date': 'Mon, 18 Dec 2023 19:42:55 GMT', 'Content-Length': '459'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Job.batch \"dagster-run-bcc712d9-a90c-4262-b8d7-f327dc74f691\" is invalid: spec.template.spec.containers[0].image: Required value","reason":"Invalid","details":{"name":"dagster-run-bcc712d9-a90c-4262-b8d7-f327dc74f691","group":"batch","kind":"Job","causes":[{"reason":"FieldValueRequired","message":"Required value","field":"spec.template.spec.containers[0].image"}]},"code":422}

```

**Comment by the4thamigo-uk** [comment]
Ok so I now understand that you have to deploy your code location via the chart https://github.com/dagster-io/dagster/tree/master/helm/dagster/charts/dagster-user-deployments.

It seems the essential difference between just deploying your own grpc server is that when you use the helm chart it adds in some extra environment variables to the pod, so that the GRPC endpoint can tell clients (such as the daemon) what docker image the endpoint it is running within. One of those environment variables in `DAGSTER_CURRENT_IMAGE`, which is literally set to the docker image repo & tag. Given this information the daemon can spin up new pods in which to execute runs, based from the same user code.

**Reply by prha** [reply, marked as answer]
Yes, that's right.

Also, for your first question, you can configure your run launcher to be the [DefaultRunLauncher](https://docs.dagster.io/_apidocs/internals#dagster._core.launcher.DefaultRunLauncher) instead of the `K8sRunLauncher`.  The default run launcher launches the run as a separate process on the GRPC server, rather than launching a k8s pod for the run.  You'd have to configure it in helm as a `customRunLauncher`.

**Reply by darktempla** [reply]
Just wondering if you can assign a different launch type per configuration? 

```yaml
load_from:
  - grpc_server:
      host: localhost
      port: 4266
      location_name: "my_grpc_server"
      launch_type: default     # for example something like this ???
```
